shotover/src/transforms/kafka/sink_cluster/mod.rs

```diff
@@ -1421,24 +1431,49 @@ impl KafkaSinkCluster {
         result
     }
 
+    /// Route broadcasted requests to all brokers split across all shotover nodes.
+    /// That is, each shotover node in a rack will deterministically be assigned a portion of the rack to route the request to.
+    /// If a shotover node is the only node in its rack it will route to all kafka brokers in the rack.
+    /// When combined with a client that is routing this request to all shotover nodes, the request will reach each kafka broker exactly once.
+    ///
+    /// The logic in this function relies on `self.shotover_nodes` and `self.nodes` being sorted by broker id.
     fn split_request_by_routing_to_all_brokers(&mut self) -> HashMap<BrokerId, ()> {
-        let mut result: HashMap<BrokerId, ()> = Default::default();
-
-        for broker in self.nodes.iter().filter(|node| {
-            node.is_up()
-                && node
-                    .rack
-                    .as_ref()
-                    .map(|rack| rack == &self.rack)
-                    // If the cluster is not using racks, include all brokers in the list.
-                    // This ensure we get full coverage of the cluster.
-                    // The client driver can filter out the resulting duplicates.
-                    .unwrap_or(true)
-        }) {
-            result.insert(broker.broker_id, ());
-        }
+        let shotovers_in_rack = self
+            .shotover_nodes
+            .iter()
+            .filter(|node| node.is_up() && node.rack == self.rack)
+            .count();
 
-        result
+        // The offset of this shotover node within the rack
+        // will be 0 <= shotover_offsets_within_rack < shotovers_in_rack
+        // The purpose is to create a deterministic way of splitting up broadcasted requests across shotover nodes.
+        let local_shotover_index_in_rack = self
+            .shotover_nodes
+            .iter()
+            .filter(|node| node.is_up() && node.rack == self.rack)
+            // start enumerating AFTER filtering down to brokers in the rack
+            .enumerate()
+            .find(|node| node.1.broker_id == self.broker_id)
+            .unwrap()
+            .0;
+
+        self.nodes
+            .iter()
+            .filter(|node| {
+                node.is_up()
+                    && node
+                        .rack
+                        .as_ref()
+                        .map(|rack| rack == &self.rack)
+                        // If the cluster is not using racks, include all brokers in the list.
+                        .unwrap_or(true)
+            })
+            // start enumerating AFTER filtering down to brokers in the rack
+            .enumerate()
+            // assign each shotover node in the rack a subset of the brokers in the rack
+            .filter(|(i, _)| i % shotovers_in_rack == local_shotover_index_in_rack)
```

Comment:
Would this panic when `shotovers_in_rack=0` (i.e., all shotover nodes in the rack are down)? This scenario could be common during OS upgrades where nodes are rolling restarted.

Reply:
shotovers_in_rack will always be at least 1, since the shotover this code is running in will always be included.
I added a comment to make this clear.